sql/src/main/java/io/crate/analyze/where/WhereClauseNullReplacer.java

```diff
@@ -0,0 +1,111 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.analyze.where;
+
+import io.crate.analyze.symbol.Function;
+import io.crate.analyze.symbol.FunctionCopyVisitor;
+import io.crate.analyze.symbol.Literal;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.execution.expression.operator.Operators;
+import io.crate.execution.expression.predicate.NotPredicate;
+import io.crate.metadata.FunctionInfo;
+
+/**
+ * Inside a where clause NULL values as logical operator arguments can be treated like boolean FALSE.
+ * As a result, the query can mostly be optimized and turned into a lucene query.
+ * Without this optimization chances are high that a genericFunctionFilter is used instead of a lucene query which
+ * results in poor performance (table scan -> filter).
+ * <p>
+ * <pre>
+ * Example:
+ *
+ *      NULL AND col1 = 1
+ *
+ *  can be handled as:
+ *
+ *      FALSE AND col1 = 1
+ *
+ *  which can be immediately normalized to:
+ *
+ *      FALSE -> NO-MATCH
+ * </pre>
+ *
+ * @implNote If a NOT predicate is encountered inside the tree, the current boolean to replace a NULL must be inverted
+ * for that leaf. Also traversing must be stopped if a conditional function is encountered as they can handle NULL
+ * values in a concrete way.
+ */
+public final class WhereClauseNullReplacer {
+
+    private static final Visitor VISITOR = new Visitor();
+
+    public static Symbol replaceIfPossible(Symbol symbol) {
+        return VISITOR.process(symbol, new Context());
+    }
+
+    private static class Context {
+        boolean insideLogicalOperator = false;
+        boolean replaceNullBoolean = false;
+    }
+
+    private static class Visitor extends FunctionCopyVisitor<Context> {
+
+        @Override
+        public Symbol visitFunction(Function func, Context context) {
+            // we won't replace anything below a conditional function as NULL are possible handled
+            if (func.info().hasFeature(FunctionInfo.Feature.CONDITIONAL)) {
+                return func;
+            }
+            boolean initialContextReplaceNullBooleanValue = context.replaceNullBoolean;
+            String functionName = func.info().ident().name();
+
+            // only operate inside logical operators
+            if (Operators.LOGICAL_OPERATORS.contains(functionName)) {
+                context.insideLogicalOperator = true;
+                if (NotPredicate.NAME.equals(functionName)) {
+                    // invert the boolean nulls are replaced with
```

Comment:
I think a comment along the lines of "`not(null)` evaluates to `null`, so rewrite `nulls` beneath `not` to `true`: `not(true) -> false`" would be more helpful instead of stating what the code does. (why vs what)